obp/ope/estimators_embed.py

```diff
@@ -0,0 +1,666 @@
+                else:
+                    theta_list.append(theta), cnf_list.append(cnf)
+                    continue
+
+            idx_list = np.argsort(cnf_list_)[::-1]
+            for idx in idx_list:
+                theta_i, cnf_i = theta_list_[idx], cnf_list_[idx]
+                theta_j, cnf_j = np.array(theta_list), np.array(cnf_list)
+                if (np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all():
+                    theta_list.append(theta_i), cnf_list.append(cnf_i)
+                else:
+                    return theta_j[-1]
+
+        return theta_j[-1]
+
+    def _estimate_with_greedy_pruning(
+        self,
+        context: np.ndarray,
+        reward: np.ndarray,
+        action: np.ndarray,
+        action_embed: np.ndarray,
+        pi_b: np.ndarray,
+        action_dist: np.ndarray,
+        position: np.ndarray,
+    ) -> float:
+        """Apply a greedy version of data-drive action embedding selection."""
+        n_emb_dim = action_embed.shape[1]
+        theta_list, cnf_list = [], []
+        current_feat, C = np.arange(n_emb_dim), np.sqrt(6) - 1
+
+        # init
+        theta, cnf = self._estimate_round_rewards(
+            context=context,
+            reward=reward,
+            action=action,
+            action_embed=action_embed[:, current_feat],
+            pi_b=pi_b,
+            action_dist=action_dist,
+            position=position,
+            with_dev=True,
+        )
+        theta_list.append(theta), cnf_list.append(cnf)
+
+        # iterate
+        while current_feat.shape[0] > self.min_emb_dim:
+            theta_list_, cnf_list_, d_list_ = [], [], []
+            for d in current_feat:
+                exclude_d_idx = np.where(current_feat != d, True, False)
+                candidate_feat = current_feat[exclude_d_idx]
+                theta, cnf = self._estimate_round_rewards(
+                    context=context,
+                    reward=reward,
+                    action=action,
+                    action_embed=action_embed[:, candidate_feat],
+                    pi_b=pi_b,
+                    action_dist=action_dist,
+                    position=position,
+                    with_dev=True,
+                )
+                d_list_.append(d)
+                theta_list_.append(theta), cnf_list_.append(cnf)
+
+            idx_list = np.argsort(cnf_list_)[::-1]
+            for idx in idx_list:
+                exclude_d = d_list_[idx]
+                theta_i, cnf_i = theta_list_[idx], cnf_list_[idx]
+                theta_j, cnf_j = np.array(theta_list), np.array(cnf_list)
+                if (np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all():
+                    theta_list.append(theta_i), cnf_list.append(cnf_i)
+                else:
+                    exclude_d_idx = np.where(current_feat != exclude_d, True, False)
+                    return theta_j[-1]
+            exclude_d_idx = np.where(current_feat != exclude_d, True, False)
```

Comment:
[ask]

This line seems to exclude a feature that has the smallest CDF.
Is this a reasonable algorithm?
(Intuitively, the smallest lower bound feature has some reason to be excluded.)

Reply:
Yes, this is intended. `theta_j[-1]` is the last element of `theta_j` (or equivalently `theta_list`), this is not necessarily the element having the smallest CDF.

Reply:
> theta_j[-1] is the last element of theta_j (or equivalently the last element of theta_list), this is not necessarily the element having the smallest CDF.

That is OK!
In my understanding, the feature extraction is implemented as the following steps:

1. The last element of `idx_list` indicate a feature that has  the smallest CDF among the `current_feature` (`idx_list = np.argsort(cnf_list_)[::-1]`)
2. The last element of `idx_list` is excluded if not returned (`idx_without_d = np.where(current_feat != excluded_dim, True, False)`)
3. That is why a feature that has the smallest CDF seems to be excluded.

Are the steps correct?
And I cannot understand why the last element of `idx_list` is excluded in the step3.

Reply:
@fullflu Thanks for clarifying your point!

> The last element of idx_list indicate a feature that has the smallest CDF among the current_feature

Let me first clarify that the last element of `idx_list` is the feature that leads to the smallest CDF **when it is excluded from OPE**. And if all features pass the criterion `(np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all()`, we know that we can exclude any feature from `current_feature` and proceed to the next round. Then, I think we should exclude the feature that leads to the smallest CDF, as it has already satisfied the criterion. (if this is not true, what you think is the appropriate feature to exclude when proceed to the next iteration?)

Reply:
Thank you for your thorough explanation!
I read the algorithm again, and I find that the process has no problem!

(I think that excluding the feature that leads to  the **smallest CDF** or the **largest CDF** might be a **greedy** algorithm, and we cannot decide which one is better.)

Reply:
@fullflu Thanks for checking again! 

> the smallest CDF or the largest CDF might be a greedy algorithm, and we cannot decide which one is better.)

I agree, but I thought the feature that leads to the smallest CDF should be removed because passing `(np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all()` becomes the most difficult with the smallest CDF, and if the feature with the smallest CDF pass the criterion, then we can be aggressive and simply remove that feature.